### PR TITLE
[WFLY-14151] Move microprofile-health extension to the microprofile feature pack

### PR DESCRIPTION
**Pull request by jmesnil**
[WFLY-14151] Move microprofile-health extension to the microprofile feature pack

This PR will handle the move of microprofile-health extension to the microprofile feature pack and the creation of a health extension to provide base healthiness features in the EE feature pack

JIRA: https://issues.redhat.com/browse/WFLY-14151

Signed-off-by: Jeff Mesnil <jmesnil@redhat.com>

### CODE REVIEW

**Comment by jmesnil** [comment]
This PR is not ready for review yet for the time being.
It's also built on top of #13756 (as the code touches a lot of similar files and I don't want to have to resolve conflicts during development of this feature).

**Comment by jmesnil** [comment]
I'm opening this PR for review. I still need to squash it before it can be merged though

**Comment by jmesnil** [comment]
This PR is built on top of #13756 (that consists of the first 3 commits).

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Health.adoc [hunk 1]
```diff
@@ -0,0 +1,87 @@
+[[Health]]
+= Health Subsystem Configuration
+
+[NOTE]
+====
+This subsystem exposes only healthiness checks for WildFly runtime.
```
Comment (bstansberry) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Health.adoc:6]: s/for WildFly/for the WildFly/g

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Health.adoc [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+The `/subsystem=health` resource defines one attribute:
+
+* `security-enabled` - a boolean to indicate whether authentication is required to access the HTTP health endpoint (described below). By default, it is `true`. The
+standalone configurations explicitly sets it to `false` to accept unauthenticated access to the HTTP endpoints.
+
+[[health-http-endpoint]]
+== HTTP Endpoint
+
+The Health HTTP endpoint is accessible on WildFly HTTP management interface http://localhost:9990/health[http://localhost:9990/health].
```
Comment (bstansberry) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Health.adoc:36]: s/on WildFly/on the WildFly/g

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Health.adoc [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+The `health` subsystem registers three HTTP endpoints:
+
+* `/health` to test both the liveness and readiness of the application server.
+* `/health/live` to test the liveness of the application server
+* `/health/ready` to test the readiness of the application server.
+
+The Health HTTP endpoints are accessible on WildFly HTTP management interface (e.g. http://localhost:9990/health[http://localhost:9990/health]).
```
Comment (bstansberry) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Health.adoc:44]: s/on WildFly/on the WildFly/g

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Health.adoc [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+[source,shell]
+----
+$ curl -v --digest -u myadminuser:myadminpassword http://localhost:9990/health
+< HTTP/1.1 200 OK
+----
+
+If the authentication fails, the  server will reply with a `401 NOT AUTHORIZED` response.
+
+NOTE: The HTTP response contains additional information with individual outcome for each probes that determined the healthiness.
```
Comment (bstansberry) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Health.adoc:78]: s/outcome/outcomes
s/probes/probe

File: ee-9/feature-pack/src/main/resources/feature_groups/domain-host-excludes.xml [hunk 5]
```diff
@@ -7,50 +7,65 @@
...
+        <param name="excluded-extensions" value="[&quot;org.wildfly.extension.clustering.web&quot;,&quot;org.wildfly.extension.health&quot;,&quot;org.wildfly.extension.metrics&quot;,&quot;org.wildfly.extension.microprofile.fault-tolerance-smallrye&quot;,&quot;org.wildfly.extension.microprofile.jwt-smallrye&quot;,&quot;org.wildfly.extension.microprofile.openapi-smallrye&quot;]"/>
     </feature>
     <feature spec="domain.host-exclude">
         <param name="host-exclude" value="WildFly16.0"/>
         <param name="host-release" value="WildFly16.0"/>
-        <param name="excluded-extensions" value="[&quot;org.wildfly.extension.microprofile.fault-tolerance-smallrye&quot;,&quot;org.wildfly.extension.microprofile.jwt-smallrye&quot;,&quot;org.wildfly.extension.microprofile.openapi-smallrye&quot;]"/>
+        <param name="excluded-extensions" value="[,&quot;org.wildfly.extension.health&quot;,&quot;org.wildfly.extension.metrics&quot;,&quot;org.wildfly.extension.microprofile.fault-tolerance-smallrye&quot;,&quot;org.wildfly.extension.microprofile.jwt-smallrye&quot;,&quot;org.wildfly.extension.microprofile.openapi-smallrye&quot;]"/>
```
Comment (bstansberry) [inline ee-9/feature-pack/src/main/resources/feature_groups/domain-host-excludes.xml:44]: First char in the list is ','

File: ee-feature-pack/galleon-common/src/main/resources/feature_groups/domain-host-excludes.xml [hunk 6]
```diff
@@ -7,35 +7,65 @@
...
         <param name="host-exclude" value="WildFly14.0"/>
         <param name="host-release" value="WildFly14.0"/>
-        <param name="excluded-extensions" value="[&quot;org.wildfly.extension.clustering.web&quot;,&quot;org.wildfly.extension.microprofile.metrics-smallrye&quot;]"/>
+        <param name="excluded-extensions" value="[&quot;org.wildfly.extension.clustering.web&quot;,&quot;org.wildfly.extension.health&quot;,&quot;org.wildfly.extension.metrics&quot;,&quot;org.wildfly.extension.microprofile.metrics-smallrye&quot;]"/>
     </feature>
     <feature spec="domain.host-exclude">
         <param name="host-exclude" value="WildFly15.0"/>
         <param name="host-release" value="WildFly15.0"/>
-        <param name="excluded-extensions" value="[&quot;org.wildfly.extension.clustering.web&quot;]"/>
+        <param name="excluded-extensions" value="[&quot;org.wildfly.extension.clustering.web&quot;,&quot;org.wildfly.extension.health&quot;,&quot;org.wildfly.extension.metrics&quot;,&quot;org.wildfly.extension.health&quot;]"/>
```
Comment (bstansberry) [inline ee-feature-pack/galleon-common/src/main/resources/feature_groups/domain-host-excludes.xml:39]: The health extension is included twice here.

File: testsuite/integration/basic/src/test/java/org/wildfly/test/integration/health/HealthHTTPEndpointTestCase.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    @Test
+    public void testHealthReadyEndpoint() throws Exception {
+        testHealthEndpoint("/health/ready");
+    }
+
+    private void testHealthEndpoint(String requestPath) throws IOException {
+        final String healthURL = "http://" + managementClient.getMgmtAddress() + ":" + managementClient.getMgmtPort() + requestPath;
+        System.out.println("healthURL = " + healthURL);
```
Comment (bstansberry) [inline testsuite/integration/basic/src/test/java/org/wildfly/test/integration/health/HealthHTTPEndpointTestCase.java:68]: ^^^

File: testsuite/integration/basic/src/test/java/org/wildfly/test/integration/health/HealthHTTPEndpointTestCase.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        try (CloseableHttpClient client = HttpClientBuilder.create().build()) {
+            CloseableHttpResponse resp = client.execute(new HttpGet(healthURL));
+            String content = EntityUtils.toString(resp.getEntity());
+            System.out.println("content = " + content);
+            System.out.println(resp.getStatusLine().getStatusCode());
+            assertEquals(200, resp.getStatusLine().getStatusCode());
```
Comment (bstansberry) [inline testsuite/integration/basic/src/test/java/org/wildfly/test/integration/health/HealthHTTPEndpointTestCase.java:74]: ^^^

File: testsuite/integration/basic/src/test/java/org/wildfly/test/integration/microprofile/opentracing/OnOffOpenTracingTestCase.java [hunk 9]
```diff
@@ -59,6 +60,7 @@
  *
  * @author Jan Stourac <jstourac@redhat.com>
  */
+@Ignore
```
Comment (bstansberry) [inline testsuite/integration/basic/src/test/java/org/wildfly/test/integration/microprofile/opentracing/OnOffOpenTracingTestCase.java:63]: Emmanuel's PR includes a JIRA number so if you add that you will avoid conflicts.

**Review by bstansberry — CHANGES_REQUESTED** [review]

File: ee-feature-pack/galleon-common/src/main/resources/feature_groups/domain-host-excludes.xml (hunk 6 above)
Comment (jmesnil) [reply ee-feature-pack/galleon-common/src/main/resources/feature_groups/domain-host-excludes.xml:39]: I really want it to be excluded! :)
sorry about that

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Health.adoc (hunk 1 above)
Comment (jmesnil) [reply docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Health.adoc]: fixed

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Health.adoc (hunk 3 above)
Comment (jmesnil) [reply docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Health.adoc]: fixed

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Health.adoc (hunk 2 above)
Comment (jmesnil) [reply docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Health.adoc]: fixed

File: ee-9/feature-pack/src/main/resources/feature_groups/domain-host-excludes.xml (hunk 5 above)
Comment (jmesnil) [reply ee-9/feature-pack/src/main/resources/feature_groups/domain-host-excludes.xml]: fixed

File: ee-feature-pack/galleon-common/src/main/resources/feature_groups/domain-host-excludes.xml (hunk 6 above)
Comment (jmesnil) [reply ee-feature-pack/galleon-common/src/main/resources/feature_groups/domain-host-excludes.xml]: fixed

File: testsuite/integration/basic/src/test/java/org/wildfly/test/integration/microprofile/opentracing/OnOffOpenTracingTestCase.java (hunk 9 above)
Comment (jmesnil) [reply testsuite/integration/basic/src/test/java/org/wildfly/test/integration/microprofile/opentracing/OnOffOpenTracingTestCase.java]: thanks, I've amended the commit in the #13756 branch too.

**Review by bstansberry — APPROVED** [review]
I'm approving this with respect to the Health change, not re other commits that come from other branches this is based upon that are handled in other PRs.

Not that I see anything wrong with those; I'm just not focusing on those commits when reviewing this PR. I've already approved #13756.

**Pull request merged**